framework-plugins/lisk-framework-chain-connector-plugin/src/chain_connector_plugin.ts

```diff
@@ -640,21 +642,35 @@ export class ChainConnectorPlugin extends BasePlugin<ChainConnectorPluginConfig>
 			await this._chainConnectorStore.setBlockHeaders(blockHeaders);
 		}
 
-		const aggregateCommits = await this._chainConnectorStore.getAggregateCommits();
-		const aggregateCommitIndex = findIndexByHeight(aggregateCommits);
-		if (aggregateCommitIndex !== -1) {
-			aggregateCommits.splice(aggregateCommitIndex, 1);
+		if (
+			!deletedBlockHeader.aggregateCommit.aggregationBits.equals(EMPTY_BYTES) ||
+			!deletedBlockHeader.aggregateCommit.certificateSignature.equals(EMPTY_BYTES)
+		) {
+			const aggregateCommits = await this._chainConnectorStore.getAggregateCommits();
+			const aggregateCommitIndex = aggregateCommits.findIndex(
+				commit => commit.height === deletedBlockHeader.aggregateCommit.height,
+			);
+			if (aggregateCommitIndex > -1) {
+				aggregateCommits.splice(aggregateCommitIndex, 1);
+			}
 			await this._chainConnectorStore.setAggregateCommits(aggregateCommits);
 		}
 
 		const validatorsHashPreimage = await this._chainConnectorStore.getValidatorsHashPreimage();
-		const validatorsHashPreimageIndex = validatorsHashPreimage.findIndex(v =>
-			v.validatorsHash.equals(newBlockHeader.validatorsHash),
-		);
-		if (validatorsHashPreimageIndex !== -1) {
-			validatorsHashPreimage.splice(validatorsHashPreimageIndex, 1);
-			await this._chainConnectorStore.setValidatorsHashPreimage(validatorsHashPreimage);
+		const activeValidatorsHashes = validatorsHashPreimage.map(b => b.validatorsHash);
+		for (const validatorHash of activeValidatorsHashes) {
+			const doesValidatorsHashExist = blockHeaders.some(b =>
+				validatorHash.equals(b.validatorsHash),
+			);
+			if (!doesValidatorsHashExist) {
+				const indexOfValidatorHash = validatorsHashPreimage.findIndex(vHash =>
+					vHash.validatorsHash.equals(validatorHash),
+				);
+				validatorsHashPreimage.splice(indexOfValidatorHash, 1);
+			}
```

Comment:
creating validatorsHash map from `blockHeaders` and filtering if it doesn't exist might be more simpler here.
ex)
```
const validatorsHashMap = blockHeaders.reduce((prev, curr) => prev[curr.validatorsHash.toString('hex')], {});
const updatedValidatorsHashPreimages = validatorsHashPreimage.filter(vhp => validatorsHashMap[vhp.validatorsHash.toString('hex')]);
if (updatedValidatorsHashPreimages.length !== validatorsHashPreimage.length) {
  await this._chainConnectorStore.setValidatorsHashPreimage(updatedValidatorsHashPreimages);
}